Libraries/Opc.Ua.Gds.Server.Common/CertificateGroup.cs

```diff
@@ -300,27 +294,128 @@ string subjectName
+                    serialNumber = authority.SerialNumber;
+                }
+                else
+                {
+                    throw new ArgumentException("Certificate does not contain an Authority Key");
+                }
+
+                if (!isCACert)
+                {
+                    if (serialNumber == certificate.SerialNumber ||
+                        X509Utils.CompareDistinguishedName(certificate.Subject, certificate.Issuer))
+                    {
+                        throw new ServiceResultException(StatusCodes.BadCertificateInvalid, "Cannot revoke self signed certificates");
+                    }
+                }
+
+                X509Certificate2 certCA = null;
+                using (ICertificateStore store = CertificateStoreIdentifier.OpenStore(storePath))
+                {
+                    if (store == null)
+                    {
+                        throw new ArgumentException("Invalid store path/type");
+                    }
+                    certCA = await X509Utils.FindIssuerCABySerialNumberAsync(store, certificate.Issuer, serialNumber);
+
+                    if (certCA == null)
+                    {
+                        throw new ServiceResultException(StatusCodes.BadCertificateInvalid, "Cannot find issuer certificate in store.");
+                    }
+
+                    if (!certCA.HasPrivateKey)
+                    {
+                        throw new ServiceResultException(StatusCodes.BadCertificateInvalid, "Issuer certificate has no private key, cannot revoke certificate.");
+                    }
+
+                    CertificateIdentifier certCAIdentifier = new CertificateIdentifier(certCA) {
+                        StorePath = storePath,
+                        StoreType = CertificateStoreIdentifier.DetermineStoreType(storePath)
+                    };
+                    X509Certificate2 certCAWithPrivateKey = await certCAIdentifier.LoadPrivateKey(issuerKeyFilePassword);
+
+                    if (certCAWithPrivateKey == null)
+                    {
+                        throw new ServiceResultException(StatusCodes.BadCertificateInvalid, "Failed to load issuer private key. Is the password correct?");
+                    }
+
+                    List<X509CRL> certCACrl = store.EnumerateCRLs(certCA, false);
+
+                    var certificateCollection = new X509Certificate2Collection() { };
+                    if (!isCACert)
+                    {
+                        certificateCollection.Add(certificate);
+                    }
+                    updatedCRL = CertificateFactory.RevokeCertificate(certCAWithPrivateKey, certCACrl, certificateCollection);
+
+                    store.AddCRL(updatedCRL);
+
+                    // delete outdated CRLs from store
+                    foreach (X509CRL caCrl in certCACrl)
+                    {
+                        store.DeleteCRL(caCrl);
+                    }
+                    store.Close();
```

Comment:
Without context I'm a little confused.

* A method called Enumerate returned a list (normally the verb Enumerate is used instead of the verb Get to indicate that it's not allocating a list a priori, but instead is just returning an iterator over existing data/data-sources)
* Unless the cert is "a CA" (which may mean self-issued, or may mean has the basic constraints CA bit set to true, I haven't seen yet) a new CRL is produced.
  * Hopefully this CRL has all the existing revocation data still in it?
* Then everything from the original list is deleted.

Since the certificate's CRL Distribution Point has a specific URL in it, why are there multiple CRLs anyways?  If you always want to delete old ones, maybe just change the method you're calling to only persist one file?

There are a lot of touch points here, is parallelism a concern?